### PR TITLE
[JENKINS-47909] Handle false hex escapes

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-47909](https://issues.jenkins-ci.org/browse/JENKINS-47909). Not sure if this is the right approach, just a proposal.

### Proposed changelog entries

* Recover from legacy user configuration folders with `$` characters that are not part of hex escape sequences.

@reviewbybees @jenkinsci/code-reviewers

### CODE REVIEW

**Review by ghost — APPROVED** [review]
:bee:

**Comment by jglick** [comment]
Deserves a migration test case added to `UserTest` I suppose.

**Review by oleg-nenashev — APPROVED** [review]
:bee: I am not sure it fully fixes the issue, but it fixes the reported defect at least.

@reviewbybees done
@olivergondza @daniel-beck @kohsuke IIUC we may need out-of-order 2.73.4 and weekly

File: core/src/main/java/jenkins/model/IdStrategy.java [hunk 1]
```diff
@@ -303,7 +303,11 @@ public String idFromFilename(@Nonnull String filename) {
                         } else {
                             break;
                         }
-                        buf.append(Character.valueOf((char)Integer.parseInt(hex.toString(), 16)));
+                        try {
+                            buf.append(Character.valueOf((char)Integer.parseInt(hex.toString(), 16)));
+                        } catch (NumberFormatException x) {
+                            buf.append('$').append(hex);
```
Comment (jglick) [inline core/src/main/java/jenkins/model/IdStrategy.java]: The main fix, to `CaseInsensitive`.

File: core/src/main/java/jenkins/model/IdStrategy.java [hunk 2]
```diff
@@ -509,7 +513,11 @@ public String idFromFilename(@Nonnull String filename) {
...
```
Comment (jglick) [inline core/src/main/java/jenkins/model/IdStrategy.java]: Actually this code long predated the security fix. But `CaseInsensitive` is the default for `SecurityRealm`; perhaps few people were using `CaseSensitive` before and thus did not happen to hit this?

File: test/src/test/java/hudson/model/UserTest.java [hunk 3]
```diff
@@ -813,6 +813,18 @@ public void emptyUsernameConfigMigrated() {
         assertThat(empty.getFullName(), equalTo("Empty"));
     }
 
+    @Issue("JENKINS-47909")
+    @LocalData
```
Comment (jglick) [inline test/src/test/java/hudson/model/UserTest.java]: Reproduces the boot failure when fix on line 306 is reverted.

**Comment by jglick** [comment]
Surprisingly, [this block of code](https://github.com/jenkinsci/jenkins/blob/e2d36820da22b734f1bce92545a950ab97d39f8d/core/src/main/java/hudson/model/User.java#L476-L508) does not get run during the migration test; the rename happens earlier, [here](https://github.com/jenkinsci/jenkins/blob/e2d36820da22b734f1bce92545a950ab97d39f8d/core/src/main/java/hudson/model/User.java#L450-L469). I suspect my comments in [JENKINS-47429](https://issues.jenkins-ci.org/browse/JENKINS-47429) provide the explanation.

**Review by ghost — APPROVED** [review]
🐝

File: test/src/test/java/hudson/model/UserTest.java [hunk 4]
```diff
@@ -813,6 +813,25 @@ public void emptyUsernameConfigMigrated() {
...
+        assertCorrectConfig(user, "users/bla$0024phem.us/config.xml");
+        assertFalse(new File(rootDir, "bla$phem.us").exists());
+        assertTrue(user.getConfigFile().getFile().exists());
+        assertThat(user.getFullName(), equalTo("Weird Username"));
+        user = User.getById("make\u1000000", false);
+        assertNotNull("we do not prevent accesses to the phony name, alas", user);
+        user = User.getById("make$1000000", false);
+        assertCorrectConfig(user, "users/make$00241000000/config.xml");
+        assertFalse(new File(rootDir, "make$1000000").exists());
+        assertTrue("but asking for the real name triggers migration", user.getConfigFile().getFile().exists());
```
Comment (jglick) [inline test/src/test/java/hudson/model/UserTest.java]: usernames

Not really fixable unless we pick a different database directory than `users`, or start storing a (usually redundant) `<id>…</id>` field, etc.

Comment (jglick) [reply test/src/test/java/hudson/model/UserTest.java]: In the first session after migration, you may see bogus usernames like this. But the config file is properly migrated as soon as you, say, log in as the real username. After the next restart, there is no trace of the bogus username `makeက000`. (Just do not get _really_ greedy and try to log in as `make$00241000000`!)

**Pull request merged**